File: python_code/line/queue/LoopQueue.py
class LoopQueue:

    def __init__(self, capacity=20):
        self.data = [0] * capacity
        self.head = self.tail = 0
        self.size = 0

    def getCapacity(self):
        return len(self.data) - 1

    def getSize(self):
        return self.size

    def enqueue(self, e):
        # 这里也可以使用 size 来判断队列是否满了，参考 issue ：https://gitee.com/douma_edu/douma-algo/issues/I4WGCE
        if (self.tail + 1) % len(self.data) == self.head:
            # 队列满了，扩容
            self.resize(self.getCapacity() * 2)
        self.data[self.tail] = e
        self.size += 1
        self.tail = (self.tail + 1) % len(self.data)

    def dequeue(self):
        if self.isEmpty():
            raise Exception("dequeue error, No Element for dequeue")
        res = self.data[self.head]
        self.data[self.head] = None
        self.size -= 1
        self.head = (self.head + 1) % len(self.data)
        if self.size == self.getCapacity() // 4:
            self.resize(self.getCapacity() // 2)
        return res

    def resize(self, new_capacity):
        new_data = [None] * (new_capacity + 1)
        for i in range(self.size):
            new_data[i] = self.data[(i + self.head) % len(self.data)]
        self.data = new_data
        self.head = 0
        self.tail = self.size

    def isEmpty(self):
        # 这里也可以使用 size 来判断队列是否空，参考 issue ：https://gitee.com/douma_edu/douma-algo/issues/I4WGCE
        return self.head == self.tail

    def getFront(self):
        if self.isEmpty():
            raise Exception("getFront error, No Element for dequeue")
        return self.data[self.head]

    def toString(self):
        res = "Queue：size = {}, capacity = {}\n".format(self.size, self.getCapacity())
        res += " 队首 ["
        i = self.head
        while i != self.tail:
            res += str(self.data[i])
            if (i + 1) % len(self.data) != self.tail:
                res += ","
            i = (i + 1) % len(self.data)
        res += "]"
        return res

File: javascript_code/line/queue/LoopQueue.js


class LoopQueue {
    constructor(capacity) {
        this.data = new Array(capacity)
        this.head = 0
        this.tail = 0
        this.size = 0
    }

    getCapacity() {
        return this.data.length - 1
    }

    getSize() {
        return this.size
    }

    enqueue(e) {
        // 这里也可以使用 size 来判断队列是否满了，参考 issue ：https://gitee.com/douma_edu/douma-algo/issues/I4WGCE
        if ((this.tail + 1) % this.data.length == this.head) {
            // 队列满了
            this.resize(this.getCapacity() * 2)
        }
        this.data[this.tail] = e
        this.size++
        this.tail = (this.tail + 1) % this.data.length
    }

    dequeue(e) {
        if (this.isEmpty()) {
            throw new Error("dequeue error, No Element for dequeue")
        }
        const res = this.data[this.head]
        this.data[this.head] = null
        this.size--
        this.head = (this.head + 1) % this.data.length
        if (this.size == Math.floor(this.getCapacity() / 4)) {
            this.resize(Math.floor(this.getCapacity() / 2))
        }
        return res
    }

    resize(newCapacity) {
        const newData = new Array(newCapacity + 1)
        for (let i = 0; i < this.size; i++) {
            newData[i] = this.data[(i + this.head) % this.data.length]
        }
        this.data = newData
        this.head = 0
        this.tail = this.size
    }

    getFront() {
        if (this.isEmpty()) {
            throw new Error("Queue is Empty")
        }
        return this.data[this.head]
    }

    isEmpty() {
        // 这里也可以使用 size 来判断队列是否空的，参考 issue ：https://gitee.com/douma_edu/douma-algo/issues/I4WGCE
        return this.head == this.tail
    }

    toString() {
        let res = `Queue：size = ${this.size}, capacity = ${this.getCapacity()}\n`
        res += " 队首 ["
        for (let i = this.head; i != this.tail; i = (i + 1) % this.data.length) {
            res += this.data[i]
            if ((i + 1) % this.data.length != this.tail) {
                res += ", "
            }
        }
        res += "]"
        return res
    }
}


const queue = new LoopQueue(10)
queue.enqueue(10)
console.log(queue.toString())

queue.enqueue(20)
console.log(queue.toString())

queue.enqueue(30)
console.log(queue.toString())

queue.dequeue()
console.log(queue.toString())

queue.dequeue()
console.log(queue.toString())